Find maximum difference between two Variables/AuxVariables

BruceWyxv opened the pull request:
### Enhancements
- Adds `ElementVariablesDifferenceMax` VectorPostprocessor class
- Adds regressions tests
- Alphabetizes VectorPostprocessor sections in Moose.C:
  - `#include`
  - `registerVectorPostprocessor()`

closes #9231

BruceWyxv committed:
Find maximum difference between two Variables/AuxVariables

- Enhancements
  - Adds `ElementVariablesDifferenceMax` VectorPostprocessor class
  - Adds regressions tests
  - Alphabetizes VPP entries (`#include` and `registerVectorPostprocessor()` blocks in Moose.C

closes #9231

moosebuild commented:
Job `Precheck:linux-gnu` on 0984980 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 7b42590210c295326eb7069e9f090a9cb4dcd5dd`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9233/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9233/style.patch | git apply -v`

moosebuild commented:
Job `Precheck:linux-gnu` on 2bc63f0 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 7b42590210c295326eb7069e9f090a9cb4dcd5dd`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9233/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9233/style.patch | git apply -v`

dschwen commented on framework/include/vectorpostprocessors/ElementVariablesDifferenceMax.h:
```diff
@@ -0,0 +1,75 @@
+#ifndef ELEMENTVARIABLESDIFFERENCEMAX_H
```
Needs a header

BruceWyxv replied:
Do I just copy over the copyright header from another file?

dschwen replied:
Yeah, from another framework file

BruceWyxv replied:
Do I update the copyright date from 2010 to 2017, or does it matter?

permcody replied:
Copy the header verbatim and leave the old date please.

BruceWyxv replied:
... hence the "DO NOT MODIFY THIS HEADER" line at the top, my bad.

dschwen commented on framework/include/vectorpostprocessors/ElementVariablesDifferenceMax.h:
```diff
@@ -0,0 +1,75 @@
+   */
+  bool _max_difference_is_negative;
+
+private:
```
I don;t see a compelling reason to make the members below private. Protected is just fine.

BruceWyxv replied:
It was an organizational choice, not design requirement. I'll move them up.

dschwen commented on framework/src/base/Moose.C:
```diff
@@ -680,23 +681,24 @@ registerObjects(Factory & factory)
   registerVectorPostprocessor(SphericalAverage);
-  registerVectorPostprocessor(CSVReader);
+  registerVectorPostprocessor(VectorOfPostprocessors);
+  registerVectorPostprocessor(VolumeHistogram);
```
Did you alphabetize those? Nice :-D

BruceWyxv replied:
You bet!

dschwen commented on framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C:
```diff
@@ -0,0 +1,143 @@
+#include "ElementVariablesDifferenceMax.h"
```
Header

dschwen commented on framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C:
```diff
@@ -0,0 +1,143 @@
+  for (unsigned int qp = 0; qp < _qrule->n_points(); ++qp)
+  {
+    // Get the difference and its sign
+    Real difference = _a[qp] - _b[qp];
```
more compact:

```
const Real difference = _furthest_from_zero ? std::abs(_a[qp] - _b[qp]) : _a[qp] - _b[qp];
```

BruceWyxv replied:
I actually had this exact line of code originally, but added some more things and replaced it. Now it makes sense to have it back in. :)

dschwen commented on framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C:
```diff
@@ -0,0 +1,143 @@
+   * is true, hence the (otherwise redundant) parenthesization of the last two
+   * terms to make a logical grouping clear to those who read this code
+   */
+  if (_preserve_sign && (_furthest_from_zero && _max_difference_is_negative))
```
this seems overly complicated. Are you sure all those sign options are needed?

BruceWyxv replied:
I believe so, as each provides useful information. I tried to make it generic so others can get the functionality they need, but want the default values to work out-of-the-box according to what most people would expect.

My specific use cases require both the largest absolute value and the largest values. The "preserve" functionality preserves information for those who care about the directionality of the difference.

The values reported by this code should be:

### All Negative
{-5, -3, -1} | Furthest from Zero: T | : F
---- | ---- | ----
Preserve Sign: T | -5 | 1
: F | 5 | 1

### Mixed, Larger Absolute in Negative
{-5, -3, -1, 3, 4} | Furthest from Zero: T | : F
---- | ---- | ----
Preserve Sign: T | -5 | 4
: F | 5 | 4

### Mixed, Large Absolute in Positive
{-4, -3, 1, 3, 5} | Furthest from Zero: T | : F
---- | ---- | ----
Preserve Sign: T | 5 | 5
: F | 5 | 5

### All Positive
{1, 3, 5} | Furthest from Zero: T | : F
---- | ---- | ----
Preserve Sign: T | 5 | 5
: F | 5 | 5

If you really are against all these options I will pull them out, as it may still be useful to others.

Otherwise, based on the "All Negative" Case I realized that it is useful to allow the combination (Furthest = F, Preserve = T), so I will change the code to allow that.

BruceWyxv replied:
If keeping all these options are acceptable and I fix the "Preserve" action
then the "All Negative" case will look like:

### All Negative
{-5, -3, -1} | Furthest from Zero: T | F
---- | ---- | ----
Preserve Sign: T | -5 | -1
: F | 5 | 1

dschwen replied:
I would just _always_ preserve the sign and if somebody wants to abs the postprocessor value they are free to do that on their own (that would keep the PP a little more focussed). But that's just my opionon.

BruceWyxv replied:
Gotcha, that makes a lot of sense.

dschwen commented on framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C:
```diff
@@ -0,0 +1,143 @@
+  gatherProxyValueMax(_max_difference[0], _max_difference_is_negative);
+  gatherProxyValueMax(_max_difference[0], _position_x[0]);
+  gatherProxyValueMax(_max_difference[0], _position_y[0]);
+  gatherProxyValueMax(_max_difference[0], _position_z[0]);
```
what are these lines supposed to do?

BruceWyxv replied:
Collect the relevant information associated with the largest difference: location and actual value in each field. I want to know more than just the largest difference, hence why this is a VectorPostprocessor instead of a regular Postprocessor.

None of the MPI sync commands work with libMesh::Point, so instead of writing an explicit version I split the coordinates into separate points.

Please let me know if there is a better method to accomplish this task!

dschwen replied:
Ok, sorry, I had to refresh myself on the `gatherProxyValueMax`. It broadcasts the second argument, for the rank that has the largest _first_ argument, to all ranks.

dschwen replied:
My coomment would be ther is no need to MPI communicate `_max_difference` and `_max_difference_is_negative`. You can infer those from _a_ and _b_.

dschwen replied:
Btw. @permcody, I think you fixed that a while ago, VPP values are per-thread, now, right? So this should be fine.

dschwen commented on test/tests/vectorpostprocessors/element_variables_difference_max/element_variables_difference_max_TT.i:
```diff
@@ -0,0 +1,89 @@
+    compare_a = u
+    compare_b = v
+    furthest_from_zero = true
+    preserve_sign = true
```
:+1: very thorough testing, but I'd prefer to have _one_ input file and then have these parameters set in the `tests` file using `cli_args`.

BruceWyxv replied:
Sounds great! First I've ever heard of this option, though. Is there existing documentation for this?

dschwen replied:
Look at [moose/test/tests/mesh/high_order_elems/tests for example](https://github.com/idaholab/moose/blob/devel/test/tests/mesh/high_order_elems/tests) for example

BruceWyxv commented:
On Sat, Jun 3, 2017 at 11:04 AM, Daniel Schwen <notifications@github.com>
wrote:

> *@dschwen* requested changes on this pull request.
> ------------------------------
>
> In framework/include/vectorpostprocessors/ElementVariablesDifferenceMax.h
> <https://github.com/idaholab/moose/pull/9233#discussion_r119990274>:
>
> > @@ -0,0 +1,75 @@
> +#ifndef ELEMENTVARIABLESDIFFERENCEMAX_H
>
> Needs a header
>
Do I just copy over the copyright header from another file?

> ------------------------------
>
> In framework/include/vectorpostprocessors/ElementVariablesDifferenceMax.h
> <https://github.com/idaholab/moose/pull/9233#discussion_r119990284>:
>
> > +  /// The x position of the associated quadrature point
> +  VectorPostprocessorValue & _position_x;
> +
> +  /// The y position of the associated quadrature point
> +  VectorPostprocessorValue & _position_y;
> +
> +  /// The z position of the associated quadrature point
> +  VectorPostprocessorValue & _position_z;
> +
> +  /**
> +   * Indicates whether #_max_difference is actually negative, used only in
> +   * finalize() if #_preserve_sign is set to true
> +   */
> +  bool _max_difference_is_negative;
> +
> +private:
>
> I don;t see a compelling reason to make the members below private.
> Protected is just fine.
>
It was an organizational choice, not design requirement. I'll move them up.

> ------------------------------
>
> In framework/src/base/Moose.C
> <https://github.com/idaholab/moose/pull/9233#discussion_r119990297>:
>
> >    registerVectorPostprocessor(SphericalAverage);
> -  registerVectorPostprocessor(CSVReader);
> +  registerVectorPostprocessor(VectorOfPostprocessors);
> +  registerVectorPostprocessor(VolumeHistogram);
>
> Did you alphabetize those? Nice :-D
>
You bet!

> ------------------------------
>
> In framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C
> <https://github.com/idaholab/moose/pull/9233#discussion_r119990300>:
>
> > @@ -0,0 +1,143 @@
> +#include "ElementVariablesDifferenceMax.h"
>
> Header
> ------------------------------
>
> In framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C
> <https://github.com/idaholab/moose/pull/9233#discussion_r119990346>:
>
> > +                     "  type = ElementVariablesDifferencesMax\n",
> +                     "\n",
> +                     "\"preserve_sign\" was set to \"true\" but \"furthest_from_zero\" is not.\n",
> +                     "This option will be ignored.\n",
> +                     "================================ ",
> +                     name(),
> +                     " ================================"));
> +}
> +
> +void
> +ElementVariablesDifferenceMax::execute()
> +{
> +  for (unsigned int qp = 0; qp < _qrule->n_points(); ++qp)
> +  {
> +    // Get the difference and its sign
> +    Real difference = _a[qp] - _b[qp];
>
> more compact:
>
> const Real difference = _furthest_from_zero ? std::abs(_a[qp] - _b[qp]) : _a[qp] - _b[qp];
>
> Got it, thanks for the detailed analysis.

I actually had this exact line of code originally, but added some more
things and replaced it. Now it makes sense to have it back in. :)

>  ------------------------------
>
> In framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C
> <https://github.com/idaholab/moose/pull/9233#discussion_r119990366>:
>
> > +  gatherProxyValueMax(_max_difference[0], _a_value[0]);
> +  gatherProxyValueMax(_max_difference[0], _b_value[0]);
> +  gatherProxyValueMax(_max_difference[0], _max_difference[0]);
> +  gatherProxyValueMax(_max_difference[0], _max_difference_is_negative);
> +  gatherProxyValueMax(_max_difference[0], _position_x[0]);
> +  gatherProxyValueMax(_max_difference[0], _position_y[0]);
> +  gatherProxyValueMax(_max_difference[0], _position_z[0]);
> +  gatherMax(_max_difference[0]);
> +
> +  /* Convert back to a negative value if required by _preserve_sign
> +   *
> +   * The sign of the maximum difference is only swapped when _furthest_from_zero
> +   * is true, hence the (otherwise redundant) parenthesization of the last two
> +   * terms to make a logical grouping clear to those who read this code
> +   */
> +  if (_preserve_sign && (_furthest_from_zero && _max_difference_is_negative))
>
> this seems overly complicated. Are you sure all those sign options are
> needed?
>
I believe so, as each provides useful information. I tried to make it
generic so others can get the functionality they need, but want the default
values to work out-of-the-box according to what most people would expect.

My specific use cases require both the largest absolute value and the
largest values. The "preserve" functionality preserves information for
those who care about the directionality of the difference.

The values reported by this code *should* be:

All Negative
{-5, -3, -1} Furthest from Zero
Preserve Sign T F
T -5 1
F 5 1
Mixed, Larger Absolute in Negative
{-5, -3, -1, 3, 4} Furthest from Zero
Preserve Sign T F
T -5 5
F 5 4
Mixed, Large Absolute in Positive
{-4, -3, 1, 3, 5} Furthest from Zero
Preserve Sign T F
T 5 5
F 5 5
All Positive
{1, 3, 5} Furthest from Zero
Preserve Sign T F
T 5 5
F 5 5



If you really are against all these options I will pull them out, as it may
still be useful to others.

Otherwise, based on the "All Negative" Case I realized that it is useful to
allow the combination (Furthest = F, Preserve = T), so I will change the
code to allow that.

> ------------------------------
>
> In framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C
> <https://github.com/idaholab/moose/pull/9233#discussion_r119990391>:
>
> > +      _position_z[0] = _q_point[qp](2);
> +    }
> +  }
> +}
> +
> +void
> +ElementVariablesDifferenceMax::finalize()
> +{
> +  // Gather all the parameters
> +  gatherProxyValueMax(_max_difference[0], _a_value[0]);
> +  gatherProxyValueMax(_max_difference[0], _b_value[0]);
> +  gatherProxyValueMax(_max_difference[0], _max_difference[0]);
> +  gatherProxyValueMax(_max_difference[0], _max_difference_is_negative);
> +  gatherProxyValueMax(_max_difference[0], _position_x[0]);
> +  gatherProxyValueMax(_max_difference[0], _position_y[0]);
> +  gatherProxyValueMax(_max_difference[0], _position_z[0]);
>
> what are these lines supposed to do?
>
Collect the relevant information associated with the largest difference:
location and actual value in each field. I want to know more than just the
largest difference, hence why this is a VectorPostprocessor instead of a
regular Postprocessor.

None of the MPI sync commands work with libMesh::Point, so instead of
writing an explicit version I split the coordinates into separate points.

Please let me know if there is a better method to accomplish this task!

> ------------------------------
>
> In test/tests/vectorpostprocessors/element_variables_difference_max/
> element_variables_difference_max_TT.i
> <https://github.com/idaholab/moose/pull/9233#discussion_r119990419>:
>
> > +
> +  [./top]
> +    type = DirichletBC
> +    variable = 'u v'
> +    boundary = 'top'
> +    value = 0
> +  [../]
> +[]
> +
> +[VectorPostprocessors]
> +  [./differenceTT]
> +    type = ElementVariablesDifferenceMax
> +    compare_a = u
> +    compare_b = v
> +    furthest_from_zero = true
> +    preserve_sign = true
>
> 👍 very thorough testing, but I'd prefer to have *one* input file and
> then have these parameters set in the tests file using cli_args.
>
Sounds great! First I've ever heard of this option, though. Is there
existing documentation for this?

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/9233#pullrequestreview-41923643>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHt6LZid-IpLflsufc53i7n0F2LLHzJnks5sAZIPgaJpZM4NvIEZ>
> .
>

BruceWyxv commented:
Sorry, this table should be:

Mixed, Larger Absolute in Negative
{-5, -3, -1, 3, 4} Furthest from Zero
Preserve Sign T F
T -5 4
F 5 4


If keeping all these options are acceptable and I fix the "Preserve" action
then the "All Negative" case will look like:

All Negative
{-5, -3, -1} Furthest from Zero
Preserve Sign T F
T -5 -1
F 5 1


On Sat, Jun 3, 2017 at 11:54 AM, Brycen Wendt <wendbryc@isu.edu> wrote:

>
>
> On Sat, Jun 3, 2017 at 11:04 AM, Daniel Schwen <notifications@github.com>
> wrote:
>
>> *@dschwen* requested changes on this pull request.
>> ------------------------------
>>
>> In framework/include/vectorpostprocessors/ElementVariablesDifferenceMax.h
>> <https://github.com/idaholab/moose/pull/9233#discussion_r119990274>:
>>
>> > @@ -0,0 +1,75 @@
>> +#ifndef ELEMENTVARIABLESDIFFERENCEMAX_H
>>
>> Needs a header
>>
> Do I just copy over the copyright header from another file?
>
>> ------------------------------
>>
>> In framework/include/vectorpostprocessors/ElementVariablesDifferenceMax.h
>> <https://github.com/idaholab/moose/pull/9233#discussion_r119990284>:
>>
>> > +  /// The x position of the associated quadrature point
>> +  VectorPostprocessorValue & _position_x;
>> +
>> +  /// The y position of the associated quadrature point
>> +  VectorPostprocessorValue & _position_y;
>> +
>> +  /// The z position of the associated quadrature point
>> +  VectorPostprocessorValue & _position_z;
>> +
>> +  /**
>> +   * Indicates whether #_max_difference is actually negative, used only in
>> +   * finalize() if #_preserve_sign is set to true
>> +   */
>> +  bool _max_difference_is_negative;
>> +
>> +private:
>>
>> I don;t see a compelling reason to make the members below private.
>> Protected is just fine.
>>
> It was an organizational choice, not design requirement. I'll move them
> up.
>
>> ------------------------------
>>
>> In framework/src/base/Moose.C
>> <https://github.com/idaholab/moose/pull/9233#discussion_r119990297>:
>>
>> >    registerVectorPostprocessor(SphericalAverage);
>> -  registerVectorPostprocessor(CSVReader);
>> +  registerVectorPostprocessor(VectorOfPostprocessors);
>> +  registerVectorPostprocessor(VolumeHistogram);
>>
>> Did you alphabetize those? Nice :-D
>>
> You bet!
>
>> ------------------------------
>>
>> In framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C
>> <https://github.com/idaholab/moose/pull/9233#discussion_r119990300>:
>>
>> > @@ -0,0 +1,143 @@
>> +#include "ElementVariablesDifferenceMax.h"
>>
>> Header
>> ------------------------------
>>
>> In framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C
>> <https://github.com/idaholab/moose/pull/9233#discussion_r119990346>:
>>
>> > +                     "  type = ElementVariablesDifferencesMax\n",
>> +                     "\n",
>> +                     "\"preserve_sign\" was set to \"true\" but \"furthest_from_zero\" is not.\n",
>> +                     "This option will be ignored.\n",
>> +                     "================================ ",
>> +                     name(),
>> +                     " ================================"));
>> +}
>> +
>> +void
>> +ElementVariablesDifferenceMax::execute()
>> +{
>> +  for (unsigned int qp = 0; qp < _qrule->n_points(); ++qp)
>> +  {
>> +    // Get the difference and its sign
>> +    Real difference = _a[qp] - _b[qp];
>>
>> more compact:
>>
>> const Real difference = _furthest_from_zero ? std::abs(_a[qp] - _b[qp]) : _a[qp] - _b[qp];
>>
>> Got it, thanks for the detailed analysis.
>
> I actually had this exact line of code originally, but added some more
> things and replaced it. Now it makes sense to have it back in. :)
>
>>  ------------------------------
>>
>> In framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C
>> <https://github.com/idaholab/moose/pull/9233#discussion_r119990366>:
>>
>> > +  gatherProxyValueMax(_max_difference[0], _a_value[0]);
>> +  gatherProxyValueMax(_max_difference[0], _b_value[0]);
>> +  gatherProxyValueMax(_max_difference[0], _max_difference[0]);
>> +  gatherProxyValueMax(_max_difference[0], _max_difference_is_negative);
>> +  gatherProxyValueMax(_max_difference[0], _position_x[0]);
>> +  gatherProxyValueMax(_max_difference[0], _position_y[0]);
>> +  gatherProxyValueMax(_max_difference[0], _position_z[0]);
>> +  gatherMax(_max_difference[0]);
>> +
>> +  /* Convert back to a negative value if required by _preserve_sign
>> +   *
>> +   * The sign of the maximum difference is only swapped when _furthest_from_zero
>> +   * is true, hence the (otherwise redundant) parenthesization of the last two
>> +   * terms to make a logical grouping clear to those who read this code
>> +   */
>> +  if (_preserve_sign && (_furthest_from_zero && _max_difference_is_negative))
>>
>> this seems overly complicated. Are you sure all those sign options are
>> needed?
>>
> I believe so, as each provides useful information. I tried to make it
> generic so others can get the functionality they need, but want the default
> values to work out-of-the-box according to what most people would expect.
>
> My specific use cases require both the largest absolute value and the
> largest values. The "preserve" functionality preserves information for
> those who care about the directionality of the difference.
>
> The values reported by this code *should* be:
>
> All Negative
> {-5, -3, -1} Furthest from Zero
> Preserve Sign T F
> T -5 1
> F 5 1
> Mixed, Larger Absolute in Negative
> {-5, -3, -1, 3, 4} Furthest from Zero
> Preserve Sign T F
> T -5 5
> F 5 4
> Mixed, Large Absolute in Positive
> {-4, -3, 1, 3, 5} Furthest from Zero
> Preserve Sign T F
> T 5 5
> F 5 5
> All Positive
> {1, 3, 5} Furthest from Zero
> Preserve Sign T F
> T 5 5
> F 5 5
>
>
>
> If you really are against all these options I will pull them out, as it
> may still be useful to others.
>
> Otherwise, based on the "All Negative" Case I realized that it is useful
> to allow the combination (Furthest = F, Preserve = T), so I will change the
> code to allow that.
>
>> ------------------------------
>>
>> In framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C
>> <https://github.com/idaholab/moose/pull/9233#discussion_r119990391>:
>>
>> > +      _position_z[0] = _q_point[qp](2);
>> +    }
>> +  }
>> +}
>> +
>> +void
>> +ElementVariablesDifferenceMax::finalize()
>> +{
>> +  // Gather all the parameters
>> +  gatherProxyValueMax(_max_difference[0], _a_value[0]);
>> +  gatherProxyValueMax(_max_difference[0], _b_value[0]);
>> +  gatherProxyValueMax(_max_difference[0], _max_difference[0]);
>> +  gatherProxyValueMax(_max_difference[0], _max_difference_is_negative);
>> +  gatherProxyValueMax(_max_difference[0], _position_x[0]);
>> +  gatherProxyValueMax(_max_difference[0], _position_y[0]);
>> +  gatherProxyValueMax(_max_difference[0], _position_z[0]);
>>
>> what are these lines supposed to do?
>>
> Collect the relevant information associated with the largest difference:
> location and actual value in each field. I want to know more than just the
> largest difference, hence why this is a VectorPostprocessor instead of a
> regular Postprocessor.
>
> None of the MPI sync commands work with libMesh::Point, so instead of
> writing an explicit version I split the coordinates into separate points.
>
> Please let me know if there is a better method to accomplish this task!
>
>> ------------------------------
>>
>> In test/tests/vectorpostprocessors/element_variables_
>> difference_max/element_variables_difference_max_TT.i
>> <https://github.com/idaholab/moose/pull/9233#discussion_r119990419>:
>>
>> > +
>> +  [./top]
>> +    type = DirichletBC
>> +    variable = 'u v'
>> +    boundary = 'top'
>> +    value = 0
>> +  [../]
>> +[]
>> +
>> +[VectorPostprocessors]
>> +  [./differenceTT]
>> +    type = ElementVariablesDifferenceMax
>> +    compare_a = u
>> +    compare_b = v
>> +    furthest_from_zero = true
>> +    preserve_sign = true
>>
>> 👍 very thorough testing, but I'd prefer to have *one* input file and
>> then have these parameters set in the tests file using cli_args.
>>
> Sounds great! First I've ever heard of this option, though. Is there
> existing documentation for this?
>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/idaholab/moose/pull/9233#pullrequestreview-41923643>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AHt6LZid-IpLflsufc53i7n0F2LLHzJnks5sAZIPgaJpZM4NvIEZ>
>> .
>>
>
>

BruceWyxv commented:
I committed some changes that should fix everything. Do I need to do something to trigger another review?

permcody commented:
No, we'll look at it. Sometimes they sit for a day or two on the weekend.
On Sat, Jun 3, 2017 at 2:22 PM BruceWyxv <notifications@github.com> wrote:

> I committed some changes that should fix everything. Do I need to do
> something to trigger another review?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/9233#issuecomment-305999221>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XIM6Uaopeu4NEYYHVdsp0wp9Nnjpgks5sAcBpgaJpZM4NvIEZ>
> .
>

moosebuild commented:
Job `Moose Documentation:linux-clang` on cd64c87 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9233/site

permcody commented on framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C:
```diff
@@ -0,0 +1,123 @@
+  gatherProxyValueMax(_max_difference[0], _b_value[0]);
+  gatherProxyValueMax(_max_difference[0], _position_x[0]);
+  gatherProxyValueMax(_max_difference[0], _position_y[0]);
+  gatherProxyValueMax(_max_difference[0], _position_z[0]);
```
Each on of these statements does two full collective communications. So that's 10 total which is pretty inefficient. I suppose this is OK for a first cut but in general you want to package up all your information and send it in a single or maybe two communications.

BruceWyxv replied:
In other words, place all the information in a single vector and gather only once?

permcody replied:
Yes, but admittedly my suggestion is optimization and as a first cut, this algorithm is probably fine. I just want you to be aware of potential problems as you scale up later. 

You might at a minimum see if you can pass a "Point" instead of breaking your position into coordinates though. It might work as many of these methods are templated and specialized to handle these common types.

BruceWyxv replied:
That is actually what I tried originally. I only broke it up when it became apparent that I would have to code some specialized MPI functions to support `Point`. This was much easier and less error-prone.

I have no problem collapsing everything into a single vector and then pulling it out later. I believe that is actually one of the supported types for the MPI/libMesh communications.

permcody commented on framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C:
```diff
@@ -0,0 +1,123 @@
+
+      _position_x[0] = _q_point[qp](0);
+      _position_y[0] = _q_point[qp](1);
+      _position_z[0] = _q_point[qp](2);
```
These three statement can be combined into one.
`_position_x[0] = _q_point[qp]`

BruceWyxv replied:
I don't understand. That is casting a `libMesh::Point` to a 'libMesh::Real`.

permcody replied:
Oh... My bad! You are correct.

permcody commented on framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C:
```diff
@@ -0,0 +1,123 @@
+  for (unsigned int qp = 0; qp < _qrule->n_points(); ++qp)
+  {
+    // Get the difference
+    const Real difference = _furthest_from_zero ? std::abs(_a[qp] - _b[qp]) : _a[qp] - _b[qp];
```
I'm confused about the logic of this calculation. Here in the execute loop "_a" and "_b" are calculated together so you are considering max differences at a single location at a time. However in the finalize loop you recalculate the max difference across processors which will allow for spatial separation between  "_a" and "_b". It's not difficult to imagine a scenario where you'll get the wrong answer because you are trapped in a local difference maxima during the execution but not when you are doing the global reduction.

BruceWyxv replied:
My understanding was that `finalize()` was run after `threadJoin()`. If true, then the calculation at the end of `finalize()` should be a global maximum.

permcody replied:
Let me see if I can paint a scenario where this algorithm goes bad. Let's say we are running on two processors and our min/max range of our _u_ and _v_ variables is [0, 1]. Now lets assume that on both processors the biggest difference we see during the execute phase between the two variables is 0.5 with values like "0.4, 0.9" and "0.2, 0.7" respectively. This is because you don't update your _a and _b values independently. Even though the fields do contain values of 0 and 1 somewhere you only consider places where they are at a max difference. Once you gather these and compute your global max, you'll have a final answer of 0.9 - 0.2 = 0.7 which isn't the global max.

BruceWyxv replied:
I'm trying but still not understanding. This scenario almost sounds like the maximum values of `_a` and `_b` are being gathered based on themselves, not `_max_difference`.

`_a` and `_b` are updated at the same time, and only when a maximum difference is found at that particular location. In the gather phase they are being collected based on the global maximum value of `_max_difference`, not their individual values.

This revision is actually based on an earlier conversation with @dschwen, when it became apparent that the maximum difference did not need to be included in the MPI communications. It could simply be calculated from the values of the two fields in `finalize()`, thus allowing a reduction in the number of MPI calls being performed.

BruceWyxv replied:
If this is regarding the failing tests in parallel, I just looked into the cause. The differences are only in the location columns and not the value columns. I believe the issue is with the symmetry of the test problem: the same `_max_difference` exists in multiple locations. The symmetry even means that the values for `_a` and `_b` are the same but the locations are different. Different mesh partitions in parallel means that a different location may be found.

In the actual use case the user will _hopefully_ be aware of the simulation symmetry and use this information to identify the region that contains the maximum differences.

I will update the tests to eliminate the symmetry.

permcody replied:
No it's not although that symmetry problem is actually a bigger issue as it's more than just symmetry that can get you. Any time you have multiple values at the extreme you are going to run into this problem so you actually need to decide on how you want to solve it. One suggestion would be to keep track of yet another piece of data such as the element or node ID and keep the one that's highest or lowest or something. Of course that also means you'll have to communicate that data in parallel since it could bounce around if all the processors don't agree on the global location of the min/max.

I'll try to illustrate again by marking up more of this code...

permcody replied:
OK, on deeper inspection, I have decided that it is the symmetry that is the problem here and that does need to be resolved. In any given problem, you could easy have the biggest difference appear in several locations in the mesh. Often it'll occur at a boundary or some other place where you are putting a forcing function so you may have a whole line of values at the max. Depending on the order in which you visit the mesh locations and the discretization your location will bounce around like crazy on different numbers of processors.

BruceWyxv replied:
I have tested from 1 to 8 tasks, and the X and Z locations are either  0.47886751345948 or 0.52113248654052. The Y locations is the same. I expect this behavior since the BCs are in Y and the ICs are symmetric in X and Z.

I dug a little deeper and noticed that the `_max_difference` for each case is actually different, just less than the tolerance of the CSVDiff. This is true for both MPI tasks and threaded cases. I added some code that performs a fuzzy equality on `_max_difference` to within `4*double_epsilon` and nothing changes. I have to increase the tolerance on the fuzzy equality to 1e-12 to start making matches. Parallel computation requires an even larger tolerance.

Should I be concerned about this? Any ideas on how to resolve this issue? I could simply create a non-symmetric problem that would pass the parallel checks, but am not sure if anyone sees a larger issue at hand that should be resolved first.

permcody replied:
I guess the bigger issue is that you may run into problems in a real production run if you change the number of processors. What exactly are you using the location for?

BruceWyxv replied:
I'm using it to identify the locations at which my functional expansion representation has the largest error. The location may be anywhere within the problem geometry.

As you had mentioned earlier: the largest error generally _is_ near boundaries. If not, however, the quality of the information is greatly improved by the specific location of the maximum error.

permcody reviewed (changes requested):
I'm not so sure this VPP belongs in the framework. We are starting to pick up a lot more of these utility type postprocessors which is good but I wonder if they don't belong in the modules somewhere. I'm sure others may disagree. Would you mind enlightening me on how you will use this in your code?

BruceWyxv commented:
Thanks everyone for the help over the weekend.

I wasn't planning on submitting this to MOOSE until @friedmud recommended it in [this moose-users discussion](https://groups.google.com/forum/#!topic/moose-users/n3Y8K9BpjKE). If it isn't going to be valuable then I can close the request. I've been learning a lot about submitting code to MOOSE, but it is also taking a lot of time.

I am using this as part of the testing for the functional expansion (FE) code. I need to evaluate the accuracy of the FE, so I have written an `AuxKernel` that samples the FE at each nodal point into an `AuxVariable`. This class is part of a suite of tools I am developing to evaluate the difference between the actual variable and its FE representation. The others tools will evaluate the average and variance of the difference.

permcody commented:
Ah! I see. So it's actually a lot like our L2/H1 error estimating postprocessors except in this case we don't have a function to evaluate so we have to compare different solution variables instead. Makes sense.

BruceWyxv commented:
Yes, exactly!

friedmud commented:
Just use maxloc() on the value. That will tell you the max value and which
processor "wins". That processor then broadcasts out the position to
everyone else.
On Mon, Jun 5, 2017 at 2:50 PM Cody Permann <notifications@github.com>
wrote:

> *@permcody* commented on this pull request.
> ------------------------------
>
> In framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C
> <https://github.com/idaholab/moose/pull/9233#discussion_r120181655>:
>
> > +  // These are all single-value arrays
> +  _a_value.resize(1);
> +  _b_value.resize(1);
> +  _max_difference.resize(1);
> +  _position_x.resize(1);
> +  _position_y.resize(1);
> +  _position_z.resize(1);
> +}
> +
> +void
> +ElementVariablesDifferenceMax::execute()
> +{
> +  for (unsigned int qp = 0; qp < _qrule->n_points(); ++qp)
> +  {
> +    // Get the difference
> +    const Real difference = _furthest_from_zero ? std::abs(_a[qp] - _b[qp]) : _a[qp] - _b[qp];
>
> OK, on deeper inspection, I have decided that it is the symmetry that is
> the problem here and that does need to be resolved. In any given problem,
> you could easy have the biggest difference appear in several locations in
> the mesh. Often it'll occur at a boundary or some other place where you are
> putting a forcing function so you may have a whole line of values at the
> max. Depending on the order in which you visit the mesh locations and the
> discretization your location will bounce around like crazy on different
> numbers of processors.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/9233#discussion_r120181655>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1JMaSeBDwTc6-X46jGPYk6KicL6iPDks5sBE4QgaJpZM4NvIEZ>
> .
>

friedmud commented:
Ok - I missed the part where you're trying to insure stability in parallel.

In general that's not going to be possible. As you spread the problem out
it will change the preconditioner and change order of operations. That will
ever so slightly change the location of the max value (if there are many
that are the "same"). The best you can do is insure that the max value and
position reported are consistent (which is what I was talking about with
maxloc()).
On Mon, Jun 5, 2017 at 4:16 PM Derek Gaston <friedmud@gmail.com> wrote:

> Just use maxloc() on the value. That will tell you the max value and which
> processor "wins". That processor then broadcasts out the position to
> everyone else.
> On Mon, Jun 5, 2017 at 2:50 PM Cody Permann <notifications@github.com>
> wrote:
>
>> *@permcody* commented on this pull request.
>> ------------------------------
>>
>> In framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C
>> <https://github.com/idaholab/moose/pull/9233#discussion_r120181655>:
>>
>> > +  // These are all single-value arrays
>> +  _a_value.resize(1);
>> +  _b_value.resize(1);
>> +  _max_difference.resize(1);
>> +  _position_x.resize(1);
>> +  _position_y.resize(1);
>> +  _position_z.resize(1);
>> +}
>> +
>> +void
>> +ElementVariablesDifferenceMax::execute()
>> +{
>> +  for (unsigned int qp = 0; qp < _qrule->n_points(); ++qp)
>> +  {
>> +    // Get the difference
>> +    const Real difference = _furthest_from_zero ? std::abs(_a[qp] - _b[qp]) : _a[qp] - _b[qp];
>>
>> OK, on deeper inspection, I have decided that it is the symmetry that is
>> the problem here and that does need to be resolved. In any given problem,
>> you could easy have the biggest difference appear in several locations in
>> the mesh. Often it'll occur at a boundary or some other place where you are
>> putting a forcing function so you may have a whole line of values at the
>> max. Depending on the order in which you visit the mesh locations and the
>> discretization your location will bounce around like crazy on different
>> numbers of processors.
>>
>> —
>> You are receiving this because you were mentioned.
>>
>>
>> Reply to this email directly, view it on GitHub
>> <https://github.com/idaholab/moose/pull/9233#discussion_r120181655>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AA1JMaSeBDwTc6-X46jGPYk6KicL6iPDks5sBE4QgaJpZM4NvIEZ>
>> .
>>
>

BruceWyxv commented:
That seems reasonable. I will change the test file so that it will (hopefully) not have the issues in parallel.

moosebuild commented:
Job `Moose Documentation:linux-clang` on fc23494 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9233/site

BruceWyxv commented:
I have tried to address all the requested changes. Please let me know if something still needs corrected or improved.

permcody reviewed (commented):
This looks pretty good to me. I guess the only remaining question is whether this should be in the framework (as opposed to one of the modules) but I'm more neutral now. Any other opinions? @idaholab/moose-team

BruceWyxv commented on framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C:
```diff
@@ -0,0 +1,126 @@
+  return params;
+}
+
+enum CollectionOfAllValuesIntoAVector
```
`CollectionOfAllValuesIntoAVector` is for easy indexing of the vector `_all`. If I use a strong-typed `enum class` then this feature is no longer implicitly available. Instead, it would require a cast each time I need to use a value.

The original intent was to provided readability and maintainability to the code.

permcody replied:
Oh yes, that's unfortunately true. The current implementation will be fine.

permcody commented:
@dschwen - any further comments?
